crates/table/src/page_pool.rs

```diff
@@ -0,0 +1,202 @@
+};
+use std::sync::Arc;
+
+/// A page pool of currently unused pages available for use in [`Pages`](super::pages::Pages).
+#[derive(Default, Clone)]
+pub struct PagePool {
+    inner: Arc<PagePoolInner>,
+}
+
+impl MemoryUsage for PagePool {
+    fn heap_usage(&self) -> usize {
+        let Self { inner } = self;
+        inner.heap_usage()
+    }
+}
+
+impl PagePool {
+    /// Puts back a [`Page`] into the pool.
+    pub fn put(&self, page: Box<Page>) {
+        self.inner.put(page);
+    }
+
+    /// Puts back a [`Page`] into the pool.
+    pub fn put_many(&self, pages: impl Iterator<Item = Box<Page>>) {
+        for page in pages {
+            self.put(page);
+        }
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports fixed rows of size `fixed_row_size`.
+    pub fn take_with_fixed_row_size(&self, fixed_row_size: Size) -> Box<Page> {
+        self.inner.take_with_fixed_row_size(fixed_row_size)
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports a maximum of `max_rows_in_page` rows.
+    fn take_with_max_row_count(&self, max_rows_in_page: usize) -> Box<Page> {
+        self.inner.take_with_max_row_count(max_rows_in_page)
+    }
+
+    /// Deserialize a page from `buf` but reuse the allocations in the pool.
+    pub fn take_deserialize_from(&self, buf: &[u8]) -> Result<Box<Page>, DecodeError> {
+        self.deserialize(bsatn::Deserializer::new(&mut &*buf))
+    }
+}
+
+impl<'de> DeserializeSeed<'de> for &PagePool {
+    type Output = Box<Page>;
+
+    fn deserialize<D: Deserializer<'de>>(self, de: D) -> Result<Self::Output, D::Error> {
+        de.deserialize_product(self)
+    }
+}
+
+impl<'de> ProductVisitor<'de> for &PagePool {
+    type Output = Box<Page>;
+
+    fn product_name(&self) -> Option<&str> {
+        Some("Page")
+    }
+
+    fn product_len(&self) -> usize {
+        2
+    }
+
+    fn visit_seq_product<A: SeqProductAccess<'de>>(self, mut prod: A) -> Result<Self::Output, A::Error> {
+        let header = prod
+            .next_element::<PageHeader>()?
+            .ok_or_else(|| A::Error::invalid_product_length(2, &self))?;
+        let row_data = prod
+            .next_element()?
+            .ok_or_else(|| A::Error::invalid_product_length(2, &self))?;
+
+        // TODO(perf, centril): reuse the allocation of `present_rows` in `page`.
+        let mut page = self.take_with_max_row_count(header.max_rows_in_page());
+        // SAFETY: `header` and `row_data` are consistent with each other.
+        unsafe { page.set_raw(header, row_data) };
+
+        Ok(page)
+    }
+
+    fn visit_named_product<A: NamedProductAccess<'de>>(self, _: A) -> Result<Self::Output, A::Error> {
+        unreachable!()
+    }
+}
+
+/// The inner actual page pool containing all the logic.
+struct PagePoolInner {
+    pages: ArrayQueue<Box<Page>>,
```

Comment:
Throwing out an idea: if this were `ArrayQueue<(Instant, Box<Page>)>`, and we stored the current time with the page when pushing it on to the queue, then age of the head of the queue would be a good metric to see if we have a larger pool than we need. 

It looks like there is no `peek` function, so we would have to update the age of the last popped page each time we pop one, but that's still doable.

If we want to try a large pool size in the future, a metric like that would be important to make sure we aren't overdoing it. It could also be used to dynamically shrink it at runtime.